### ISSUE TITLE
Improve BM3D speed

### ISSUE THREAD

**Issue by kvttt** [deepinv/deepinv#1194]
### Describe the feature

Currently, `deepinv.models.BM3D` relies on `deepinv.models.utils.tensor2array` and `deepinv.models.utils.array2tensor` to transfer between PyTorch and NumPy. This is because `bm3d.bm3d` runs on the CPU and requires the input to be `numpy.ndarray`. There are two main consequences as a result of this. 
1. When the tensor is originally on the GPU, doing so requires transferring between the GPU and the CPU, which introduces overhead. The effect of this overhead is more significant when BM3D is used as a plug-and-play prior since it is repeatedly called at each iteration.
2. The algorithm behind BM3D is well-suited for parallelization since each "reference patch" is processed independently. The original `bm3d.bm3d` already takes advantage of parallelization but further acceleration is possible by porting everything to the GPU. I have recently re-implemented BM3D using CuPy (see [bm3d-cupy](https://github.com/kvttt/bm3d_cupy)). I also included a benchmarking script so that one can compare and see the speedup on their own machine. The only dependencies are PyWavelets and CuPy. However, the original `bm3d` contains hard-coded wavelets, whereas all wavelets I used are from PyWavelets, so the results might not be fully reproducible. Nevertheless, the performance is similar and speedup is ~16x on my machine (see below). The actual speedup is potentially even higher if the transferring overhead is taken into consideration. 

<img width="3818" height="763" alt="Image" src="https://github.com/user-attachments/assets/a6c31920-84a5-4917-9eac-9c48f8e6b71e" />

```
BM3D (CPU) time: 3.0155 (0.0226) s
  PSNR: 30.85 dB
  SSIM: 0.8194


BM3D (GPU) time: 0.1881 (0.0004) s
  Speedup: 16.03x
  PSNR: 30.90 dB
  SSIM: 0.8228
```

Following @jscanvic's suggestion, I would like to ask the community for opinions. To start this off, here are some of my initial thoughts.

Implementation-wise, there are two options:
- Option 1: Port my code (written with CuPy) to PyTorch. 
- Option 2: Keep my code in CuPy and add helper functions to handle shape and zero-copy conversion between CuPy and PyTorch.

As I mentioned above, my re-implementation uses wavelets from PyWavelets whereas the original `bm3d` hard-codes some of the wavelets. In order to enforce consistency between the wavelets that are used, it is possible to use the function `bm4d._get_transf_matrix` to get the exact wavelets that the original `bm3d` and `bm4d` are using. One benefit of using `bm4d._get_transf_matrix` is better reproducibility. However, it also makes `bm4d` a dependency.

On the user-facing side, my initial thought is to have a keyword argument like `use_legacy=True` to keep the old behavior (use `bm3d.bm3d`) for backward compatibility. The user can then set this keyword argument to `False` to use the GPU-accelerated version.

### Motivation

To have a faster BM3D.

### Additional context

Request for this new feature is originally brought up in #1192.

**Comment by mh-nguyen712** [comment]
Thanks for raising this! 
One up vote for porting your code to pytorch, which makes it more compatible with the rest of the library! 
The `use_legacy` is a good idea. 
I'll be happy to review!

**Comment by jscanvic** [comment]
Thank you for the great write-up!

I'm happy with both options & think it would be a great addition to the library!

+1 for the backwards compatibility bit too

**Comment by kvttt** [comment]
Update: I replace CuPy functions with the corresponding PyTorch functions line-by-line. It turns out using PyTorch instead of CuPy is even faster (see below). The dependency on `cupy` can be dropped but still requires the discrete cosine transform from `scipy` and wavelets from `pywt`.

<img width="4780" height="764" alt="Image" src="https://github.com/user-attachments/assets/6ed6e16a-c8af-461d-945d-387cdb08f5ac" />

```
BM3D (CPU) time: 2.8493 (0.0349) s
  PSNR: 30.85 dB
  SSIM: 0.8194


BM3D (CuPy) time: 0.1950 (0.0013) s
  Speedup: 14.61x
  PSNR: 30.90 dB
  SSIM: 0.8237


BM3D (Torch) time: 0.1277 (0.0007) s
  Speedup: 22.31x
  PSNR: 30.90 dB
  SSIM: 0.8237
```

**Comment by Tmodrzyk** [comment]
I didn't know BM3D was CPU only ! Thanks for the issue and for the fix, I'm for using PyTorch for the implementation. 😄

**Comment by romainvo** [comment]
Super nice! agreed on `use_legacy`, and +1 for your `torch` implementation, even more if it's faster :)!

**Comment by kvttt** [comment]
I create a PR at #1195 and I would like to ask for some additional feedback.

1. I dropped dependency on `pywt` and used `ptwt` instead. It turns out that `ptwt.wavedec` does not support `mode="periodization"`, which is the mode that I use to generate the wavelet transform matrices. I do have a short workaround to make sure `ptwt.wavedec`'s behavior match as close as possible to that of `pywt.wavedec`. But the workaround can only guarantee reproducibility up to flipping the sign and switching the rows around. In my opinion this discrepancy is benign. I do think using `ptwt` rather than `pywt` is good because the former is PyTorch-native.
2. Any ideas on what kind of new tests I need to include? I'm thinking maybe having one test for the old `bm3d.bm3d`, one test for the new `bm3d_fast` on the CPU, and one test for the new `bm3d_fast` on the GPU. Maybe the `bm3d_fast` on the CPU test is unnecessary but I'm not sure.

**Comment by mh-nguyen712** [comment]
Hi @kvttt, thanks for the PR. 
1. It seems to me that `ptwt` and `pywt` are compatible (and we use both of them in the library). I'm not sure if I fully understand what you meant, will look closer to your PR. 
2. It would be great to test that `use_legacy = False` matches the output of old bm3d, and `bm3d_fast` produces good output: eg close to old bm3d with a tolerence and good PSNR (see `test_denoiser_perf` in test_models. Don't hesitate if you have other ideas.

**Comment by kvttt** [comment]
Hi @mh-nguyen712, 

> It seems to me that `ptwt` and `pywt` are compatible (and we use both of them in the library). I'm not sure if I fully understand what you meant, will look closer to your PR. 

I thought they are compatible too. However, there is one functionality that `pywt` has but `ptwt` does not. `pywt.wavedec` supports `mode='periodization'`, which pads periodically and output the least amount of coefficients. However, `ptwt.wavedec` does not support this mode. It only supports `mode='periodic'`. This is fine because all the coefficients we need are already there, except that they are repeated and some coefficients have their signs flipped. The following snippet of code demonstrate this problem. Case 1 demonstrates the behavior of `pywt`. It gives the least amount of coefficients. The coefficients can then be easily stacked into a column of the wavelet transform matrix. Case 2 demonstrates the behavior of `ptwt`. At each level, all the coefficients are produced. Some are repeated and some have their signs flipped. Case 3 demonstrates my workaround, i.e., manually truncating the coefficients.

```
import numpy as np
import torch
import pywt
import ptwt

np.set_printoptions(precision=2)
torch.set_printoptions(precision=2)

# Case 1
I = np.eye(8)
basis = I[:, 0]
coeffs = pywt.wavedec(basis, 'bior1.5', mode='periodization', level=3)
for coeff in coeffs:
    print(coeff)

# Case 2
I = torch.eye(8)
basis = I[:, 0]
coeffs = ptwt.wavedec(basis, 'bior1.5', mode='periodic', level=3)
for coeff in coeffs:
    print(coeff)

# Case 3
coeffs = [coeff[: 2 ** max(0, j - 1)] for j, coeff in enumerate(coeffs)]
for coeff in coeffs:
    print(coeff)
```

Output:
```
[0.35]
[0.23]
[ 0.59 -0.09]
[0.71 0.   0.   0.  ]

tensor([0.35, 0.35, 0.35, 0.35, 0.35, 0.35, 0.35, 0.35])
tensor([-0.23, -0.23, -0.23, -0.23, -0.23, -0.23, -0.23, -0.23])
tensor([-0.09,  0.59, -0.09,  0.59, -0.09,  0.59, -0.09,  0.59])
tensor([0.00, 0.00, 0.71, 0.00, 0.00, 0.00, 0.71, 0.00])

tensor([0.35])
tensor([-0.23])
tensor([-0.09,  0.59])
tensor([0.00, 0.00, 0.71, 0.00])
```
